internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,393 @@
+
+			if err := r.Client.Update(ctx, capiCluster); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+	}
+
+	log = log.WithValues("cluster", capiCluster.Name)
+
+	// Wait for controlplane to be ready. This should never be false as the predicates
+	// do the filtering.
+	if !capiCluster.Status.ControlPlaneReady && !conditions.IsTrue(capiCluster, clusterv1.ControlPlaneReadyCondition) {
+		log.Info("clusters control plane is not ready, requeue")
+		return ctrl.Result{RequeueAfter: defaultRequeueDuration}, nil
+	}
+
+	// Collect errors as an aggregate to return together after all patches have been performed.
+	var errs []error
+
+	result, err := r.reconcile(ctx, capiCluster)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error reconciling cluster: %w", err))
+	}
+
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		capiClusterCopy := capiCluster.DeepCopy()
+
+		patchBase := client.MergeFromWithOptions(capiCluster, client.MergeFromWithOptimisticLock{})
+
+		if err := r.Client.Patch(ctx, capiClusterCopy, patchBase); err != nil {
+			errs = append(errs, fmt.Errorf("failed to patch cluster: %w", err))
+		}
+		return nil
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(errs) > 0 {
+		return ctrl.Result{}, errorutils.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcile(ctx context.Context, capiCluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// placeholder name/namespace must be set to be able to test this but it won't be used.
+	rancherCluster := &managementv3.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "c-",
+			Namespace: capiCluster.Namespace,
+			Labels: map[string]string{
+				capiClusterOwner:          capiCluster.Name,
+				capiClusterOwnerNamespace: capiCluster.Namespace,
+			},
+		},
+	}
+
+	rancherClusterList := &managementv3.ClusterList{}
+	selectors := []client.ListOption{
+		client.MatchingLabels{
+			capiClusterOwner:          capiCluster.Name,
+			capiClusterOwnerNamespace: capiCluster.Namespace,
+			ownedLabelName:            "",
+		},
+	}
+	err := r.RancherClient.List(ctx, rancherClusterList, selectors...)
+
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, fmt.Sprintf("Unable to fetch rancher cluster %s", client.ObjectKeyFromObject(rancherCluster)))
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if len(rancherClusterList.Items) != 0 {
+		if len(rancherClusterList.Items) > 1 {
+			log.Info("More than one rancher cluster found. Will default to using the first one.")
+		}
+
+		rancherCluster = &rancherClusterList.Items[0]
+	}
+
+	if !capiCluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.deleteDependentRancherCluster(ctx, capiCluster); err != nil {
+			return ctrl.Result{Requeue: true}, fmt.Errorf("error deleting associated managementv3.Cluster resources: %w", err)
```

Comment:
```suggestion
			return ctrl.Result{}, fmt.Errorf("error deleting associated managementv3.Cluster resources: %w", err)
```
Error causes automatic requeue.

Reply:
If I'm not mistaken in case of error `Requeue: true` will be ignored. Controller runtime will handle exponential backoff when error is returned.